### PR TITLE
clusteroperator: Report when OLM reaches "level" and check syncs

### PR DESCRIPTION
**Pull request by smarterclayton**
Cluster operators are expected to report the version of the payload
they are included in once they are "deployed", and also to keep the
cluster operator object created. Have the OLM operator keep CO up
to date, report the payload version once it hits available, and use
the count of successful syncs from the queueInformers as a probalistic
measurement of "available" (i.e. is the operator able to retire syncs).
A future change should add a "health over time" metric or a "has
successfully synced all InstallPlans at least once" metric to replace
the current estimation.

### CODE REVIEW

**Comment by smarterclayton** [comment]
Looks like a legit failure

/lgtm cancel

**Comment by smarterclayton** [comment]
```
--- FAIL: TestCSVCopyWatchingAllNamespaces (220.56s)
	operator_groups_e2e_test.go:1129: Creating CRD
	operator_groups_e2e_test.go:1136: Getting default operator group 'global-operators' installed via operatorgroup-default.yaml openshift-operators
	operator_groups_e2e_test.go:1144: Waiting on operator group to have correct status
	operator_groups_e2e_test.go:1159: Creating CSV
	operator_groups_e2e_test.go:1229: wait for CSV to succeed
	csv_e2e_test.go:226: Pending (RequirementsUnknown): requirements not yet checked
	csv_e2e_test.go:226: Installing (InstallSucceeded): waiting for install components to report healthy
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Installing (InstallWaiting): installing: Waiting: waiting for deployment operator-deploymentsjjqt to become ready: Waiting for rollout to finish: 0 of 2 updated replicas are available...
	csv_e2e_test.go:226: Succeeded (InstallSucceeded): install strategy completed with no errors
	operator_groups_e2e_test.go:1233: Waiting for operator namespace csv to have annotations
	operator_groups_e2e_test.go:1252: Found CSV count of 32
	operator_groups_e2e_test.go:1254: Create other namespace
	operator_groups_e2e_test.go:1268: Waiting to ensure copied CSV shows up in other namespace
	operator_groups_e2e_test.go:1286: Modifying operator group to no longer watch all namespaces
	require.go:794: 
			Error Trace:	operator_groups_e2e_test.go:1312
			Error:      	Received unexpected error:
			            	timed out waiting for the condition
			Test:       	TestCSVCopyWatchingAllNamespaces
	operator_groups_e2e_test.go:1293: Re-modifying operator group to be watching all namespaces
```

```
time="2019-03-12T13:43:20Z" level=info msg="updated annotations to match current operatorgroup" csv=another-csv-t8j44 id=o7RZJ namespace=openshift-operators phase=Pending
time="2019-03-12T13:43:20Z" level=info msg="requirements were not met" csv=another-csv-t8j44 id=o7RZJ namespace=openshift-operators phase=Pending
time="2019-03-12T13:43:21Z" level=info msg="retrying openshift-operators/another-csv-t8j44"
E0312 13:43:21.247565       1 queueinformer_operator.go:175] Sync "openshift-operators/another-csv-t8j44" failed: requirements were not met
time="2019-03-12T13:43:21Z" level=info msg="checking b-hlwmw-stable"
time="2019-03-12T13:43:21Z" level=info msg="checking another-csv-t8j44"
time="2019-03-12T13:43:21Z" level=info msg="csv in operatorgroup" csv=another-csv-t8j44 id=4OXdF namespace=openshift-operators opgroup=global-operators phase=Pending
time="2019-03-12T13:43:21Z" level=info msg="updated annotations to match current operatorgroup" csv=another-csv-t8j44 id=4OXdF namespace=openshift-operators phase=Pending
time="2019-03-12T13:43:21Z" level=info msg="requirements were not met" csv=another-csv-t8j44 id=4OXdF namespace=openshift-operators phase=Pending
time="2019-03-12T13:43:21Z" level=info msg="retrying openshift-operators/another-csv-t8j44"
E0312 13:43:21.641556       1 queueinformer_operator.go:175] Sync "openshift-operators/another-csv-t8j44" failed: error transitioning ClusterServiceVersion: requirements were not met and error updating CSV status: error updating ClusterServiceVersion status: Operation cannot be fulfilled on clusterserviceversions.operators.coreos.com "another-csv-t8j44": the object has been modified; please apply your changes to the latest version and try again
time="2019-03-12T13:43:21Z" level=info msg="checking b-hlwmw-stable"
time="2019-03-12T13:43:21Z" level=info msg="checking another-csv-t8j44"
time="2019-03-12T13:43:21Z" level=info msg="csv in operatorgroup" csv=another-csv-t8j44 id=dCSyy namespace=openshift-operators opgroup=global-operators phase=Pending
time="2019-03-12T13:43:21Z" level=info msg="updated annotations to match current operatorgroup" csv=another-csv-t8j44 id=dCSyy namespace=openshift-operators phase=Pending
time="2019-03-12T13:43:21Z" level=info msg="requirements were not met" csv=another-csv-t8j44 id=dCSyy namespace=openshift-operators phase=Pending
time="2019-03-12T13:43:22Z" level=info msg="retrying openshift-operators/another-csv-t8j44"
E0312 13:43:22.095308       1 queueinformer_operator.go:175] Sync "openshift-operators/another-csv-t8j44" failed: requirements were not met
time="2019-03-12T13:43:22Z" level=info msg="checking another-csv-t8j44"
time="2019-03-12T13:43:22Z" level=info msg="checking b-hlwmw-stable"
```

File: cmd/olm/main.go [hunk 1]
```diff
@@ -128,98 +128,198 @@ func main() {
...
-				log.Info("Existing operator status not found, creating")
-				created, err := configClient.ClusterOperators().Create(&configv1.ClusterOperator{
-					ObjectMeta: metav1.ObjectMeta{
-						Name: *writeStatusName,
-					},
-				})
-				if err != nil {
-					log.Fatalf("ClusterOperator create failed: %v\n", err)
+
+		// Sample the sync channel and see whether we're successfully retiring syncs as a
```
Comment (alecmerdler) [inline cmd/olm/main.go]: Did you mean "retrying", not "retiring"?

Comment (smarterclayton) [reply cmd/olm/main.go]: No, retiring.  Term for when you complete something and don't have to deal with it anymore - you "retire a transaction from a retry loop"

**Comment by smarterclayton** [comment]
This is really wierd, the error is consistent in this PR.  Trying to spot what I changed - I might have slowed the sync loop down such that we're now flaking more (would be my first guess).  Have we seen this test flake before?

**Comment by smarterclayton** [comment]
Hrm, error also happened in https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/operator-framework_operator-lifecycle-manager/750/pull-ci-operator-framework-operator-lifecycle-manager-master-e2e-aws-olm/1225/

So I *think* that flake is not this PR, but this PR might cause it to happen more frequently.

**Comment by njhale** [comment]
I have a pretty good idea of what's causing the flakes with `TestUpdateCSVWithOwnedAPIService` at least. I should have a PR with a fix by sometime tomorrow morning.

**Comment by njhale** [comment]
#758 opened - should help out some.

**Pull request merged**